sound/soc/sof/loader.c

```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
 	return ext_man_size;
 }
 
+static int snd_sof_fw_ext_man4_parse(struct snd_sof_dev *sdev, const struct firmware *fw)
+{
+	struct sof_ipc4_fw_modules *module_entry;
+	struct sof_man4_fw_binary_header *fw_header;
+	struct sof_ext_manifest4_hdr *hdr;
+	struct sof_man4_module_config *fm_config;
+	struct sof_man4_module_entry *fm_entry;
+	int fw_offset;
+	int i;
+
+	if (fw->size < sizeof(hdr)) {
+		dev_err(sdev->dev, "Invalid fw size\n");
+		return -EINVAL;
+	}
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+
+	fw_offset = hdr->len;
+
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + fw_offset +
+							 SOF_MAN4_FW_HDR_OFFSET);
+	dev_dbg(sdev->dev, " fw %s: header length %x, module num %d", fw_header->name,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module_entry *)((void *)fw_header + fw_header->len);
+	fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);
+
+	sdev->fw_modules = devm_kmalloc_array(sdev->dev, fw_header->num_module_entries,
+					      sizeof(*module_entry), GFP_KERNEL);
+	if (!sdev->fw_modules)
+		return -ENOMEM;
+
+	sdev->fw_module_num = fw_header->num_module_entries;
+	module_entry = sdev->fw_modules;
```

Comment:
wow the variable names and struct names are so confusing. You  have variable fw_entry pointing to a struct named sof_man4_module_entry. And you have a variable module_entry pointing to a struct name sof_ipc4_fw_modules. I dont know how to decipher this. I trust this is doing the right thing but so hard to read and understand what these are used for.

Reply:
> wow the variable names and struct names are so confusing. You have variable fw_entry pointing to a struct named sof_man4_module_entry. And you have a variable module_entry pointing to a struct name sof_ipc4_fw_modules. I dont know how to decipher this. I trust this is doing the right thing but so hard to read and understand what these are used for.

fw_entry is pointing to the data of firmware binary, and we copy some items to module_entry (sdev->fw_modules) which will be used by SOF. @RanderWang Please add your comments if I miss something.

Reply:
yes, Here we don't exactly copy the module config in FW manifest, but create another structure to make it more simple and general (some flags maybe , then copy some of them. I also discussed with bard about this. @ranj063 do you prefer to fully save module config ?

Reply:
> there are 2 separate issues. I was literally only talking about making the variable names more intuitive for readers to understand whats going on. But on the other issue of whether ew should copy and save only certain modules depends on how scalable that is. Why do we do this today? Is it because SOF doesnt support all modules? What happens if and when we support all modules? Will we still need to do this?

Currently, we only store what we really need in `struct sof_ipc4_fw_modules {}`, but I don't know if we will need more in the future. Maybe we can copy the entire `sof_man4_module_entry` so we don't need to add new variables in the future